From 22bde562330bbad6149a6e6729a2490404dfdb43 Mon Sep 17 00:00:00 2001
From: Frajo Haider <f_haider@gmx.at>
Date: Tue, 28 Jan 2020 13:39:19 +0000
Subject: [PATCH 14/18] Revert "tcg: Add INDEX_op_dupm_vec"

This reverts commit 37ee55a081b7863ffab2151068dd1b2f11376914.

Conflicts:
	tcg/tcg-op-gvec.c
---
 tcg/aarch64/tcg-target.inc.c |  4 ---
 tcg/i386/tcg-target.inc.c    |  4 ---
 tcg/tcg-op-gvec.c            | 86 ++++++++++++++++++++++----------------------
 tcg/tcg-op-vec.c             | 11 ------
 tcg/tcg-op.h                 |  1 -
 tcg/tcg-opc.h                |  1 -
 tcg/tcg.c                    |  1 -
 7 files changed, 42 insertions(+), 66 deletions(-)

diff --git a/tcg/aarch64/tcg-target.inc.c b/tcg/aarch64/tcg-target.inc.c
index 3f92101..f669338 100644
--- a/tcg/aarch64/tcg-target.inc.c
+++ b/tcg/aarch64/tcg-target.inc.c
@@ -2306,9 +2306,6 @@ static void tcg_out_vec_op(TCGContext *s, TCGOpcode opc,
     case INDEX_op_st_vec:
         tcg_out_st(s, type, a0, a1, a2);
         break;
-    case INDEX_op_dupm_vec:
-        tcg_out_dupm_vec(s, type, vece, a0, a1, a2);
-        break;
     case INDEX_op_add_vec:
         tcg_out_insn(s, 3616, ADD, is_q, vece, a0, a1, a2);
         break;
@@ -2737,7 +2734,6 @@ static const TCGTargetOpDef *tcg_target_op_def(TCGOpcode op)
         return &w_w;
     case INDEX_op_ld_vec:
     case INDEX_op_st_vec:
-    case INDEX_op_dupm_vec:
         return &w_r;
     case INDEX_op_dup_vec:
         return &w_wr;
diff --git a/tcg/i386/tcg-target.inc.c b/tcg/i386/tcg-target.inc.c
index c7ead57..97dbc96 100644
--- a/tcg/i386/tcg-target.inc.c
+++ b/tcg/i386/tcg-target.inc.c
@@ -2886,9 +2886,6 @@ static void tcg_out_vec_op(TCGContext *s, TCGOpcode opc,
     case INDEX_op_st_vec:
         tcg_out_st(s, type, a0, a1, a2);
         break;
-    case INDEX_op_dupm_vec:
-        tcg_out_dupm_vec(s, type, vece, a0, a1, a2);
-        break;
 
     case INDEX_op_x86_shufps_vec:
         insn = OPC_SHUFPS;
@@ -3175,7 +3172,6 @@ static const TCGTargetOpDef *tcg_target_op_def(TCGOpcode op)
 
     case INDEX_op_ld_vec:
     case INDEX_op_st_vec:
-    case INDEX_op_dupm_vec:
         return &x_r;
 
     case INDEX_op_add_vec:
diff --git a/tcg/tcg-op-gvec.c b/tcg/tcg-op-gvec.c
index 3707c0effb..3984f01697 100644
--- a/tcg/tcg-op-gvec.c
+++ b/tcg/tcg-op-gvec.c
@@ -459,53 +459,6 @@ static TCGType choose_vector_type(const TCGOpcode *list, unsigned vece,
     return 0;
 }
 
-static void do_dup_store(TCGType type, uint32_t dofs, uint32_t oprsz,
-                         uint32_t maxsz, TCGv_vec t_vec)
-{
-    uint32_t i = 0;
-
-    tcg_debug_assert(oprsz >= 8);
-
-    /*
-     * This may be expand_clr for the tail of an operation, e.g.
-     * oprsz == 8 && maxsz == 64.  The first 8 bytes of this store
-     * are misaligned wrt the maximum vector size, so do that first.
-     */
-    if (dofs & 8) {
-        tcg_gen_stl_vec(t_vec, cpu_env, dofs + i, TCG_TYPE_V64);
-        i += 8;
-    }
-
-    switch (type) {
-    case TCG_TYPE_V256:
-        /*
-         * Recall that ARM SVE allows vector sizes that are not a
-         * power of 2, but always a multiple of 16.  The intent is
-         * that e.g. size == 80 would be expanded with 2x32 + 1x16.
-         */
-        for (; i + 32 <= oprsz; i += 32) {
-            tcg_gen_stl_vec(t_vec, cpu_env, dofs + i, TCG_TYPE_V256);
-        }
-        /* fallthru */
-    case TCG_TYPE_V128:
-        for (; i + 16 <= oprsz; i += 16) {
-            tcg_gen_stl_vec(t_vec, cpu_env, dofs + i, TCG_TYPE_V128);
-        }
-        break;
-    case TCG_TYPE_V64:
-        for (; i < oprsz; i += 8) {
-            tcg_gen_stl_vec(t_vec, cpu_env, dofs + i, TCG_TYPE_V64);
-        }
-        break;
-    default:
-        g_assert_not_reached();
-    }
-
-    if (oprsz < maxsz) {
-        expand_clr(dofs + oprsz, maxsz - oprsz);
-    }
-}
-
 /* Set OPRSZ bytes at DOFS to replications of IN_32, IN_64 or IN_C.
  * Only one of IN_32 or IN_64 may be set;
  * IN_C is used if IN_32 and IN_64 are unset.
@@ -545,11 +498,49 @@ static void do_dup(unsigned vece, uint32_t dofs, uint32_t oprsz,
         } else if (in_64) {
             tcg_gen_dup_i64_vec(vece, t_vec, in_64);
         } else {
-            tcg_gen_dupi_vec(vece, t_vec, in_c);
+            switch (vece) {
+            case MO_8:
+                tcg_gen_dup8i_vec(t_vec, in_c);
+                break;
+            case MO_16:
+                tcg_gen_dup16i_vec(t_vec, in_c);
+                break;
+            case MO_32:
+                tcg_gen_dup32i_vec(t_vec, in_c);
+                break;
+            default:
+                tcg_gen_dup64i_vec(t_vec, in_c);
+                break;
+            }
+        }
+
+        i = 0;
+        switch (type) {
+        case TCG_TYPE_V256:
+            /* Recall that ARM SVE allows vector sizes that are not a
+             * power of 2, but always a multiple of 16.  The intent is
+             * that e.g. size == 80 would be expanded with 2x32 + 1x16.
+             */
+            for (; i + 32 <= oprsz; i += 32) {
+                tcg_gen_stl_vec(t_vec, cpu_env, dofs + i, TCG_TYPE_V256);
+            }
+            /* fallthru */
+        case TCG_TYPE_V128:
+            for (; i + 16 <= oprsz; i += 16) {
+                tcg_gen_stl_vec(t_vec, cpu_env, dofs + i, TCG_TYPE_V128);
+            }
+            break;
+        case TCG_TYPE_V64:
+            for (; i < oprsz; i += 8) {
+                tcg_gen_stl_vec(t_vec, cpu_env, dofs + i, TCG_TYPE_V64);
+            }
+            break;
+        default:
+            g_assert_not_reached();
         }
-        do_dup_store(type, dofs, oprsz, maxsz, t_vec);
+
         tcg_temp_free_vec(t_vec);
-        return;
+        goto done;
     }
 
     /* Otherwise, inline with an integer type, unless "large".  */
@@ -1544,12 +1535,7 @@ void tcg_gen_gvec_dup_mem(unsigned vece, uint32_t dofs, uint32_t aofs,
     check_size_align(oprsz, maxsz, dofs);
     if (vece <= MO_64) {
         TCGType type = choose_vector_type(NULL, vece, oprsz, 0);
-        if (type != 0) {
-            TCGv_vec t_vec = tcg_temp_new_vec(type);
-            tcg_gen_dup_mem_vec(vece, t_vec, cpu_env, aofs);
-            do_dup_store(type, dofs, oprsz, maxsz, t_vec);
-            tcg_temp_free_vec(t_vec);
-        } else if (vece <= MO_32) {
+        if (vece <= MO_32) {
             TCGv_i32 in = tcg_temp_new_i32();
             switch (vece) {
             case MO_8:
diff --git a/tcg/tcg-op-vec.c b/tcg/tcg-op-vec.c
index 6714991..d5e4097 100644
--- a/tcg/tcg-op-vec.c
+++ b/tcg/tcg-op-vec.c
@@ -312,17 +312,6 @@ void tcg_gen_dup_i32_vec(unsigned vece, TCGv_vec r, TCGv_i32 a)
     vec_gen_2(INDEX_op_dup_vec, type, vece, ri, ai);
 }
 
-void tcg_gen_dup_mem_vec(unsigned vece, TCGv_vec r, TCGv_ptr b,
-                         tcg_target_long ofs)
-{
-    TCGArg ri = tcgv_vec_arg(r);
-    TCGArg bi = tcgv_ptr_arg(b);
-    TCGTemp *rt = arg_temp(ri);
-    TCGType type = rt->base_type;
-
-    vec_gen_3(INDEX_op_dupm_vec, type, vece, ri, bi, ofs);
-}
-
 static void vec_gen_ldst(TCGOpcode opc, TCGv_vec r, TCGv_ptr b, TCGArg o)
 {
     TCGArg ri = tcgv_vec_arg(r);
-- 
1.8.3-rc3

